### PR TITLE
Load grid data

### PR DESCRIPTION
**Pull request by TomNicholas**
Progress towards solving #17 .

You can now load a grid file alongside the dump files, and they will be merged together.

Including a grid file is optional, and you can also load just the grid file on it's own if you want using the new function `open_grid()`.

If it finds `coordinates_type='orthogonal'` then it will rename dimensions, add Orthogonal Toroidal Coordinates and add `Rxy` and `Zxy` as Cylindrical coordinates. I'm unsure how this should work in general - should there be flags/auto-detection of simpler geometries? i.e. special-case s-alpha, slab etc? 

Also refactored to separate code into different modules more cleanly.

I named the coordinates such that when you open a general grid file you get a dataset like:
```
<xarray.Dataset>
Dimensions:       (theta: 64, x: 100)
Coordinates:
    R             (x, theta) float64 ...
    Z             (x, theta) float64 ...
    psi           (x, theta) float64 ...
  * theta         (theta) float64 0.0 0.09973 0.1995 ... 6.084 6.183 6.283
Dimensions without coordinates: x
Data variables:
    dx            (x, theta) float64 ...
    dy            (x, theta) float64 ...
    ShiftAngle    (x) float64 ...
    zShift        (x, theta) float64 ...
    pol_angle     (x, theta) float64 ...
    ShiftTorsion  (x, theta) float64 ...
    Gxy           (x, theta) float64 ...
    Hxy           (x, theta) float64 ...
    etaxy         (x, theta) float64 ...
    beta          (x, theta) float64 ...
    yshift        (x, theta) float64 ...
    Bpxy          (x, theta) float64 ...
    Btxy          (x, theta) float64 ...
    Bxy           (x, theta) float64 ...
    hthe          (x, theta) float64 ...
    hrad          (x, theta) float64 ...
    sinty         (x, theta) float64 ...
    thetaxy       (x, theta) float64 ...
    yxy           (x, theta) float64 ...
    pressure      (x, theta) float64 ...
    Jpar0         (x, theta) float64 ...
    Ni0           (x, theta) float64 ...
    Te0           (x, theta) float64 ...
    Ti0           (x, theta) float64 ...
    bxcvx         (x, theta) float64 ...
    bxcvy         (x, theta) float64 ...
    bxcvz         (x, theta) float64 ...
    g11           (x, theta) float64 ...
    g22           (x, theta) float64 ...
    g33           (x, theta) float64 ...
    g12           (x, theta) float64 ...
    g13           (x, theta) float64 ...
    g23           (x, theta) float64 ...
    g_11          (x, theta) float64 ...
    g_22          (x, theta) float64 ...
    g_33          (x, theta) float64 ...
    g_12          (x, theta) float64 ...
    g_13          (x, theta) float64 ...
    g_23          (x, theta) float64 ...
    J             (x, theta) float64 ...
Attributes:
    coordinates_type:   orthogonal
    git_hash:           38fb7f181bf01f1e6c83aed9e31ad76d2159a218
    BOUT_version:       [4 2 2]
    Hypnotoad_version:  [1 2 2]
    grid:               {'nx': 100, 'ny': 64, 'y_boundary_guards': 0, 'ixseps...
```

### CODE REVIEW

**Comment by ZedThree** [comment]
`orthogonal` coordinates, i.e. the identity transform, doesn't necessarily mean toroidal coordinates -- they could be slab, a straight cylinder, a 2D plane, etc.

It's definitely a cool feature, but it should need something more explicit to enable it.

**Comment by TomNicholas** [comment]
Okay, how about an attribute like `{'geometry': 'toroidal'}`?

Then we could have logic like 

1) If `'geometry'` supplied as keyword arg to `open_boutdataset()` or `open_grid()` then use that

2) If no kwarg supplied then look for it in global attributes of dump/grid files

3) If still not found then either assume a geometry or throw a warning?

Then we could have an easy way for people to add their own coordinate systems, specified by new possible values of `'geometry'`?

**Comment by ZedThree** [comment]
That sounds reasonable, though the name might need to be bikeshed-ed a little.

If it's not found, then just do nothing, or warn and have a "quiet" option.

**Comment by johnomotani** [comment]
The labels in `coordinates_type` are horrible at the moment... see https://github.com/boutproject/BOUT-dev/issues/1747. `coordinates_type == orthogonal` is actually for `ShiftedMetric`, i.e. toroidal coordinates.

**Comment by TomNicholas** [comment]
So I made choosing and defining types of geometry as explicit as possible, by having the coordinate adjustments defined by functions, and you choose from a dictionary of such functions by name, e.g. `'toroidal'`, `'s-alpha'`. The desired geometry type is passed as a keyword argument to `open_grid()` or `open_boutdataset`, following the behaviour @ZedThree suggested.

This approach is easily extendable because these "geometry types" are registered to a global dictionary of possible geometries (`REGISTERED_GEOMETRIES`) using the `@register_geometry` decorator. So if a user wants to define a new geometry type they don't have to submit a PR, they can just use the decorator on their own function definition and it will become an option available locally.

For example, imagine I (somehow) wrote a BOUT module that does simulations of atmospheric circulation on the whole Earth. I might want to convert the dimensions `('x', 'z')` to 1D coordinates `('lat', 'lon')`. All I have to do is write a function which does that and use the decorator to register it under the desired name:

```python
@register_geometry('geographic')
def add_geographic_coords(ds):

    # Add longitude
    ds = ds.rename(x='lon', inplace=True)
    lon = xr.DataArray(np.linspace(start=0, stop=2*np.pi, num=ds.dims['lon']), dims='lon')
    ds = ds.assign_coords(lon=lon)

    # Add latitude
    ds = ds.rename(z='lat', inplace=True)
    lat = xr.DataArray(np.linspace(start=-np.pi/2, stop=np.pi/2, num=ds.dims['lat']), dims='lat')
    ds = ds.assign_coords(lat=lat)

    return ds
```

After this code is run then `'geographic'` will be a newly available type of geometry:
```python
ds = open_boutdataset('BOUT.dmp.*.nc', gridfilepath='grid.nc', geometry='geographic')
```

Geometries which might be useful for lots of people should go in the new `geometries.py`, as all geometries in there will be registered on importing xBOUT.

If geometries end up being a more complex concept then just a single function applied once then this approach should be reimplemented as registering geometry classes which override certain methods of a `GeometryBase` class.

**Comment by ZedThree** [comment]
That definitely sounds more like the right thing to do. The decorator to register the new geometries is very cool

File: xbout/geometries.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+    Raises
+    ------
+    UnregisteredGeometryError
+    """
+
+    try:
+        add_geometry_coords = REGISTERED_GEOMETRIES[geometry_name]
+    except KeyError:
+        raise UnregisteredGeometryError
```
Comment (ZedThree) [inline xbout/geometries.py]: I think you can pass a string here to make the error message friendlier. As you've subclassed from `exception`, it should just do the right thing

File: xbout/grid.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+    gridfilepath = Path(gridfilepath)
+    grid = xr.open_dataset(gridfilepath, engine=_check_filetype(gridfilepath))
+
+    # TODO find out what 'yup_xsplit' etc are in the doublenull storm file John gave me
+    # For now drop any variables with extra dimensions
+    variables = list(grid.variables)
+    vars_to_drop = [var for var in variables
+                    if not all(dim in ['t', 'x', 'y', 'z']
+                               for dim in grid[var].dims)]
```
Comment (ZedThree) [inline xbout/grid.py]: This is clever, but unfortunately difficult to read! Please could you rewrite this to be a bit simpler?

File: xbout/geometries.py (hunk 1 above)
Comment (TomNicholas) [reply xbout/geometries.py]: Good idea!

**Comment by TomNicholas** [comment]
The tests are failing because it's trying to use scipy to write out a netcdf file, but scipy isn't in our list of dependencies so it's not installed on the test server. That's despite me specifying `engine='netcdf4'` to `.to_netcdf()`.

**Comment by johnomotani** [comment]
I found the cause of the test failures - `tmpdir_factory.mktemp()` returns some kind of `LocalPath` object. `xarray`'s `to_netcdf` expects the path to be a `str`, if the `path` argument is not either `None` or a `str`, `xarray` assumes it is a 'file-like' object and forces the engine to be `'scipy'` to handle it https://github.com/pydata/xarray/blob/38ee23f96035df95ee3baf06645ef5a21616f47f/xarray/backends/api.py#L844-L845

Also `Path(filepath)` fails in Python-3.5 if `filepath` is a `LocalPath` object, although it works in Python-3.6.

I've pushed a couple of commits that fix the test failures.

**Comment by TomNicholas** [comment]
Oh nice catch @johnomotani! Okay I'll merge now

**Pull request merged**